src/search/plugins/raw_registry.cc

```diff
@@ -1,56 +1,197 @@
 #include "raw_registry.h"
 
+#include "plugin.h"
+#include "registry.h"
+
+#include "../utils/collections.h"
+#include "../utils/strings.h"
+
+#include <string>
+
 using namespace std;
 
 namespace plugins {
-RawPluginInfo::RawPluginInfo(
-    const string &key,
-    const Any &factory,
-    const string &group,
-    const PluginTypeNameGetter &type_name_factory,
-    const DocFactory &doc_factory,
-    const type_index &type)
-    : key(key),
-      factory(factory),
-      group(group),
-      type_name_factory(type_name_factory),
-      doc_factory(doc_factory),
-      type(type) {
+void RawRegistry::insert_category_plugin(const CategoryPlugin &category_plugin) {
+    category_plugins.push_back(&category_plugin);
 }
 
-void RawRegistry::insert_plugin_type_data(
-    type_index type, const string &type_name, const string &documentation,
-    const string &predefinition_key, const string &alias,
-    const PredefinitionFunction &predefinition_function) {
-    plugin_types.emplace_back(type, type_name, documentation, predefinition_key,
-                              alias, predefinition_function);
+void RawRegistry::insert_subcategory_plugin(const SubcategoryPlugin &subcategory_plugin) {
+    subcategory_plugins.push_back(&subcategory_plugin);
 }
 
-void RawRegistry::insert_plugin_group_data(
-    const string &group_id, const string &doc_title) {
-    plugin_groups.emplace_back(group_id, doc_title);
+void RawRegistry::insert_enum_plugin(const EnumPlugin &enum_plugin) {
+    enum_plugins.push_back(&enum_plugin);
 }
 
-void RawRegistry::insert_plugin_data(
-    const string &key,
-    const Any &factory,
-    const string &group,
-    PluginTypeNameGetter &type_name_factory,
-    DocFactory &doc_factory,
-    type_index &type) {
-    plugins.emplace_back(key, factory, group, type_name_factory, doc_factory,
-                         type);
+void RawRegistry::insert_plugin(const Plugin &plugin) {
+    plugins.push_back(&plugin);
```

Comment:
If I see it correctly `insert_plugin` is called whenever a plugin object is created. Could we also determine here whether `plugin` is also, e.g.,  an `EnumPlugin` (probably via `dynamic_cast`) and if yes add it to `enum_plugins` instead of calling the functions `insert_enum_plugin`, ... explicitly? It probably wouldn't be a better solution anyway but this also seems a little strange to me :).

Reply:
The different plugins (`Plugin`, `EnumPlugin`, `CategoryPlugin`, `SubcategoryPlugin`) do not derive from a common base class, so we cannot pass them to the same function. We could add a common base class but then we would give up the knowledge about what kind of thing we are registering on the calling side, only to recover that knowledge with a cast inside the function. Maybe I'm not understanding your suggestion completely, but so far, I don't see an advantage.

Reply:
Ah right, I somehow thought they were more closely related. Even if they were, your second point answers my question and it makes sense that dropping the information just to recompute it is not a great idea.